File: 22. Generate Parentheses/generateParenthesis.java
class generateParenthesis {
    List<String> res = new ArrayList<String>();
    public List<String> solution(int n) {
        dfs(n, n, "");
        return res;
    }

    public void dfs(int left, int right, String str) {
        if (left == 0 && right == 0) {
            res.add(str);
            return;
        }
        if (left > 0) {
            dfs(left - 1, right, str + "(");
        }
        if (right > left) {
            dfs(left, right - 1, str + ")");
        }
    }
}

File: 22. Generate Parentheses/generateParenthesis.py
class generateParenthesis:
    def __init__(self):
        self.res = []
        
    def solution(self, n: int):
        self.dfs(n, n, "")
        return self.res
    
    def dfs(self, left, right, str):
        if left == 0 and right == 0:
            self.res.append(str)
            return
        if left > 0:
            self.dfs(left - 1, right, str + "(")
        if right > left:
            self.dfs(left, right - 1, str + ")")

obj = generateParenthesis()
res = obj.solution(3)
print(res)